src/rust/inetstack/protocols/udp/peer.rs

```diff
@@ -234,126 +263,126 @@ impl<const N: usize> UdpPeer<N> {
-        UdpPopFuture::new(queue.get_recv_queue(), size)
+
+        // Check that the queue still exists.
+        let mut queue: SharedUdpQueue<N> = match self.get_shared_queue(&qd) {
+            Ok(queue) => queue.clone(),
+            Err(e) => return (qd, OperationResult::Failed(e)),
+        };
+        let yielder: Yielder = Yielder::new();
+        let result: Result<(SocketAddrV4, DemiBuffer), Fail> = queue.pop(size, yielder).await;
+        match result {
+            Ok((addr, buf)) => (qd, OperationResult::Pop(Some(addr), buf)),
+            Err(e) => (qd, OperationResult::Failed(e)),
+        }
     }
 
     /// Consumes the payload from a buffer.
-    pub fn do_receive(&mut self, ipv4_hdr: &Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
+    pub fn receive(&mut self, ipv4_hdr: &Ipv4Header, buf: DemiBuffer) -> Result<(), Fail> {
         #[cfg(feature = "profiler")]
         timer!("udp::receive");
-        let qtable: Ref<IoQueueTable> = self.qtable.borrow();
         // Parse datagram.
         let (hdr, data): (UdpHeader, DemiBuffer) = UdpHeader::parse(ipv4_hdr, buf, self.checksum_offload)?;
         debug!("UDP received {:?}", hdr);
 
         let local: SocketAddrV4 = SocketAddrV4::new(ipv4_hdr.get_dest_addr(), hdr.dest_port());
         let remote: SocketAddrV4 = SocketAddrV4::new(ipv4_hdr.get_src_addr(), hdr.src_port());
 
-        let recv_queue: SharedQueue<SharedQueueSlot<DemiBuffer>> = match self.bound.get(&local) {
-            Some(qd) => match qtable.get::<UdpQueue>(&qd) {
-                Ok(queue) => queue.get_recv_queue(),
-                Err(_) => return Err(Fail::new(libc::ENOTCONN, "port not bound")),
-            },
+        let mut queue: SharedUdpQueue<N> = match self.addr_in_use(&local) {
+            Some(queue) => queue,
             None => {
                 // Handle wildcard address.
                 let local: SocketAddrV4 = SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, hdr.dest_port());
-                match self.bound.get(&local) {
-                    Some(qd) => match qtable.get::<UdpQueue>(&qd) {
-                        Ok(queue) => queue.get_recv_queue(),
-                        Err(_) => return Err(Fail::new(libc::ENOTCONN, "port not bound")),
-                    },
-                    // TODO: Send ICMPv4 error in this condition.
+                match self.addr_in_use(&local) {
+                    Some(queue) => queue,
                     None => return Err(Fail::new(libc::ENOTCONN, "port not bound")),
                 }
             },
         };
         // TODO: Drop this packet if local address/port pair is not bound.
+        queue.receive(remote, data)
+    }
 
-        // Push data to the receiver-side shared queue. This will cause the
-        // associated pool operation to be ready.
-        recv_queue.push(SharedQueueSlot { local, remote, data }).unwrap();
+    fn addr_in_use(&self, local: &SocketAddrV4) -> Option<SharedUdpQueue<N>> {
```

Comment:
How about adding the `get_` prefix to match the naming style to other get* functions?